src/components/donation/Steps/DonateMethods/Stripe/Form.tsx

```diff
@@ -0,0 +1,166 @@
+import { yupResolver } from "@hookform/resolvers/yup";
+import { useEffect, useState } from "react";
+import { FormProvider, useForm } from "react-hook-form";
+import { Link } from "react-router-dom";
+import { ObjectSchema, number, object } from "yup";
+import { FormValues, Props } from "./types";
+import { SchemaShape } from "schemas/types";
+import { useStripeCurrenciesQuery } from "services/apes";
+import { useErrorContext } from "contexts/ErrorContext";
+import { Currency } from "components/CurrencySelector";
+import ExtLink from "components/ExtLink";
+import LoadText from "components/LoadText";
+import QueryLoader from "components/QueryLoader";
+import Split from "components/Split";
+import { CheckField, Field } from "components/form";
+import { requiredString } from "schemas/string";
+import { GENERIC_ERROR_MESSAGE } from "constants/common";
+import { appRoutes } from "constants/routes";
+import { TERMS_OF_USE_DONOR } from "constants/urls";
+import AdvancedOptions from "../../../AdvancedOptions";
+import CurrencySelectorField from "./CurrencySelectorField";
+
+type FormProps = Props & {
+  defaultValues: Partial<FormValues>;
+  onSubmit: (formValues: FormValues) => void | Promise<void>;
+};
+
+export default function Form(props: FormProps) {
+  const [countryCode, setCountryCode] = useState<string>();
+
+  const queryState = useStripeCurrenciesQuery(
+    { country_code: countryCode || "" },
+    { skip: !countryCode }
+  );
+  const { handleError } = useErrorContext();
+
+  useEffect(() => {
+    fetch("https://ipapi.co/json/")
+      .then<{ country_code: string }>((response) => response.json())
+      .then((data) => setCountryCode(data.country_code))
+      .catch((err) => handleError(err, GENERIC_ERROR_MESSAGE));
+  });
+
+  return (
+    <QueryLoader
+      queryState={queryState}
+      classes={{ container: "grid justify-center" }}
+    >
+      {(data) => (
+        <Content {...props} currencies={data.supported_payment_currencies} />
+      )}
+    </QueryLoader>
+  );
+}
+
+function Content({
+  advanceOptDisplay,
+  currencies,
+  defaultValues,
+  state: { recipient },
+  widgetConfig,
+  onSubmit,
+}: FormProps & { currencies: string[] }) {
+  const methods = useForm<FormValues>({
+    resolver: yupResolver(schema),
+    defaultValues: {
+      amount: defaultValues.amount,
+      currency: { code: currencies[0].toUpperCase() },
+      pctLiquidSplit: defaultValues.pctLiquidSplit ?? 50,
+      userOptForKYC: recipient.isKYCRequired || defaultValues.userOptForKYC, // if KYC required, user opts in by default
+    },
+  });
+
+  const isInsideWidget = widgetConfig !== null;
+
+  const submit = methods.handleSubmit(onSubmit);
+
+  return (
+    <FormProvider {...methods}>
+      <form onSubmit={submit} className="grid gap-4">
+        <Field<FormValues>
```

Comment:
I tried to donate `$0.01` and gets error - should add min amount validation

<img width="1281" alt="image" src="https://github.com/AngelProtocolFinance/angelprotocol-web-app/assets/89639563/8f5a74fa-6061-4e2d-8e68-59a85feda008">

Reply:
Good catch.

I'd agree that the error modal can be made more clear/helpful by showing this specific error. And although validating a field even before the submission is made would be even better, there is a drawback, namely that we need to get an exchange rate dynamically.

To do this we need to get the current exchange rate between whatever currency was loaded for the users location and USD. Stripe has no API that would allow reading such a value, as they depend on 3rd parties for this data, so we'd have to access some 3rd party as well. 
I couldn't find a free API that would serve our purposes, only payed ones and these were the most often recommended ones:
- https://exchangeratesapi.io - pricing goes from $10/month for 10,000 API calls per month to $80/month for 500,000 API calls per month
- https://openexchangerates.org - $12/month for 10,000 API calls -> $97/month for unlimited API calls

@SovereignAndrey @ap-justin @jp-mariano do you know of any other APIs?

@SovereignAndrey so these are the 2 options for handling the problem:
1. After submit show the error modal stating "Amount must be at least $0.50 usd"
   - pros: cheap, easy to implement
   - cons: no field validation (better said, postponed field validation), the same error would show for every currency requiring the user to look up the current exchange rate themselves before inputting a value
2. Use a 3rd party API to get the current exchange rate
   - pros: field validation, minimum amount can be tailored to the selected currency ("Amount must be at least 51 RSD" or "...156 GRD)
   - cons: paid service

Reply:
@misicnenad If we tweak the setup and make the assumption that we don't nesessarily need live rates for checkout, this becomes more manageable. 

I'd suggest we use OpenExchangeRates.org (herein: OER) for a source of FX rates. We can do up to 1000 requests per month for their free tier. After that, it's $10/month for up to 10K requests. 

Periodically, an APES AWS Lambda (via a 2-3x day EventBridge trigger) would hit the OER API and update a DB table with the latest rates.

We would have an AWS endpoint for the web app to query the most recent FX rate for a given currency from AWS DB. This rate would be recent enough I think for most cases.

At 3x a day for 10 currencies, that's under the 1000 / month free tier limit! I think we can get away with 2x a day, since fiat FX rates do not move nearly as much as crypto. Wdyt?

Reply:
We could use Wise for forex rates too. For reference, https://docs.wise.com/api-docs/api-reference/rate

Reply:
This would work. It would be even better if we were querying the exchange rates, but storing the minimum donation amounts (MDA) in this DB table. That way our webapp can simply check and validate the amount the user has input in the donation field.

> We would have an AWS endpoint for the web app to query the most recent FX rate for a given currency from AWS DB. This rate would be recent enough I think for most cases.
> 
> At 3x a day for 10 currencies...

No need to query for multiple currencies, we only need the current exchange rate for USD which returns *exchange rates for all currencies against USD* (see https://docs.openexchangerates.org/reference/api-introduction#api-response-formats).
We could query 3x/day only against USD, use the exchange rates to recalculate MDA and store those values in a table, so the webapp can query MDA against this table for any possible currency.

Edge cases:
1. the client wants to donate in a currency for which we do not have a minimum amount as _openexchangerates_ didn't provide us with the exchange rate.
   - **solution:** we stop using _openexchangerates_ as they should support all possible currencies; we switch to another provider.
2. the exchange rate changes sufficiently enough for a given currency to cause the "Amount must be at least $0.50 usd" error to be thrown in certain cases, until our lambda updates MDA table; this might not even be a black swan event, just a natural change of the rate.
   - **solution:** for the error to appear in such cases would require a user to want to donate the bare minimum in their currency, which is not going to happen that often anyway. If it does then it would be understandable to show the error to users. We might implement certain steps to handle this though, as it would be bad UX to keep showing this error to all users for a couple of hours, before our lambda updates the rates table.
   Potential steps:
      - **set MDA to always be a bit higher than the $0.50 rate** -> if calculate the current exact MDA for PHP to be 27.85 (~$0.50), we increase it by 10% just to be sure and store 30 in our DB; we do this for every currency. This measure should be enough to avoid MDA errors as the exchange rate rarely changes 10% or more in a single day (even Venezuelan Bolivar doesn't change this drastically); we could increase the value by even less, maybe 5%.
      NOTE: we could round this value down, as showing a message like "Mimimum donation amount is 30" is much more pleasant than "Mimimum donation amount is 30.7615".
      - **if a user experiences the "Amount must be at least $0.50 usd" error has occurred, we automatically trigger the MDA updater lambda** -> if upon submitting the donation we see that the error above has occurred, this must mean there was a wild swing in exchange rate(s) and that we should immediately trigger the MDA updater lambda to update the MDA table; we do this so that future potential donors **do not** experience the same error. Despite this, the MDA updater lambda continues triggering as scheduled.
      
NOTE:  **we should add a message directly on screen or as part of an info popup explaining why MDA might be different than the last time they were donating** -> users might find it confusing to see that the minimum donation amount is 30 when the day before it was 29; adding a message explaining the relationship between MDA and the current exchange rate would help them understand that MDA didn't just change because we're greedy.

Reply:
@SovereignAndrey

Reply:
> if we were querying the exchange rates, but storing the minimum donation amounts (MDA) in this DB table

@misicnenad If we store __both__ items, then we can validate MDA is hit and also can show an FX rate in UI (if it becomes important in the future).

Reply:
> @SovereignAndrey , @misicnenad We could use Wise for forex rates too. For reference, https://docs.wise.com/api-docs/api-reference/rate

@jp-mariano missed this comment as GH doesn't update comment threads in real-time, only on reload.

I think this solves our problem then. We could still have a special lambda that updates the MDA table just to make fetching MDA faster. Then we also update [stripe-currencies](https://us-east-1.console.aws.amazon.com/lambda/home?region=us-east-1#/functions/stripe-currencies?tab=code) to fetch MDAs for all supported currencies based on user's locale and voila.

@SovereignAndrey 
> If we store both items, then we can validate MDA is hit and also can show an FX rate in UI (if it becomes important in the future).
> ...
> Ex: The minimum donation amount is 1 USD or 420.69 NZD (calculated based on a rate of 841.38 NZD/USD).

We can do that.

> We could start with US$1 as our internal MDA for fiat...

Sounds good :+1:

Reply:
@SovereignAndrey 
> If we store both items, then we can validate MDA is hit and also can show an FX rate in UI (if it becomes important in the future).
...
Ex: The minimum donation amount is 1 USD or 420.69 NZD (calculated based on a rate of 841.38 NZD/USD).

Take a look at the current tooltip for the minimum donation amount:
![image](https://github.com/AngelProtocolFinance/angelprotocol-web-app/assets/19427053/dae4eaf5-4c1c-4368-bae4-c2c286bfd47c)

Do you want me to add the `"... (calculated based on a rate of 107 RSD/USD)"` part as well?

---
Currently we store rates with 2 decimals, but minimum donation amount with rounded to the nearest integer (so 106.91 RSD/USD becomes 107 minimum donation amount). Do you want us to store the minimum donation amounts with 2 decimals as well?

Reply:
@misicnenad 
> Take a look at the current tooltip for the minimum donation amount:

Looks great! No need to add the other words.

> Do you want us to store the minimum donation amounts with 2 decimals as well?

Agreed! We prob should just use the nearest whole number.